File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

class Solution {
    private void dfs(
        List<String> results,
        Map<String, PriorityQueue<String>> fromToMap,
        String from
    ) {
        while (fromToMap.containsKey(from) && !fromToMap.get(from).isEmpty()) {
            dfs(results, fromToMap, fromToMap.get(from).poll());
        }
        
        results.add(0, from);
    }
    
    public String[] solution(String[][] tickets) {
        List<String> results = new LinkedList<>();
        Map<String, PriorityQueue<String>> fromToMap = new HashMap<>();
        
        for (String[] ticket : tickets) {
            fromToMap.putIfAbsent(ticket[0], new PriorityQueue<>());
            fromToMap.get(ticket[0]).add(ticket[1]);
        }
        
        dfs(results, fromToMap, "ICN");
        
        return results.toArray(new String[0]);
    }
}

File: 프로그래머스/3/43164. 여행경로/여행경로.kt
import java.util.PriorityQueue

class Solution {
    fun solution(tickets: Array<Array<String>>): Array<String> {
        val results = mutableListOf<String>()
        val fromToMap = mutableMapOf<String, PriorityQueue<String>>()
        
        tickets.forEach { (from, to) ->
            fromToMap.getOrPut(from) { PriorityQueue() }.add(to)
        }
        
        fun dfs(from: String) {
            val destinations = fromToMap[from]
            while (destinations != null && destinations.isNotEmpty()) {
                dfs(destinations.poll())
            }
            results.add(0, from)
        }
        
        dfs("ICN")
        
        return results.toTypedArray()
    }
}